ports/stm32/boards/make-af-csv.py

```diff
@@ -0,0 +1,253 @@
+    heading = ["Port", "Pin"] + af_domains
+    col_width = [4] * len(heading)
+    # Second csv row lists the peripherals handled by that column
+    category_parts = [defaultdict(set) for _ in range(len(heading))]
+    # Third row onwards are the individual pin function mappings
+    rows = []
+    for pin in sorted(mapping, key=pin_int):
+        functions = mapping[pin]
+        row = [""] * len(heading)
+        row[0] = "Port%s" % pin[1]
+        row[1] = pin
+        for signal, af in functions:
+            # Some signals are loose about whether 1 is declared or not.
+            signal = signal.replace("CAN_", "CAN1_")
+            signal = signal.replace("I2S_", "I2S1_")
+            if signal.startswith("ETH"):
+                signal = eth_remap(signal)
+            column = heading.index(re.search("(AF\d+)", af).group(1))
+            peripheral = re.search("([A-Z1-9]*[A-Z]+)(\d*)(_|$)", signal)
+            category_parts[column][peripheral[1]].add(peripheral[2])
+            signal = signal.replace("SYS_", "")
+            signal = signal.replace("-", "/")
+            if row[column]:
+                if signal not in row[column].split("/"):
+                    # multiple af signals per pin
+                    row[column] = "/".join((row[column], signal))
+            else:
+                row[column] = signal
+
+        if pin in adc_pins:
+            row[-1] = "/".join(adc_pins[pin])
+
+        for i, val in enumerate(row):
+            col_width[i] = max(col_width[i], len(val))
+
+        rows.append(row)
+
+    # Simplify the category sets, eg TIM1/TIM2 -> TIM1/2
+    category_head = ["" for _ in range(len(heading))]
+    for i, periph in enumerate(category_parts):
+        periphs = []
+        for key, vals in periph.items():
+            if len(vals) > 1:
+                if "" in vals:
+                    vals.remove("")
+                    vals.add("1")
+                periphs.append(key + "/".join(sorted(vals, key=int)))
+            else:
+                periphs.append(key + vals.pop())
+
+        category_head[i] = "/".join(sorted(periphs))
+        col_width[i] = max(col_width[i], len(category_head[i]))
+    category_head[-1] = "ADC"
+
+    # Sort the rows by the first two columns; port and pin number
+    rows.sort(key=lambda row: (row[0], pin_int(row[1])))
+
+    output_file = Path(__file__).parent / ("%s_af.csv" % target.lower())
+    with output_file.open("w") as out:
+        padded_heading = [val.ljust(col_width[i]) for i, val in enumerate(heading)]
+        print(",".join(padded_heading), file=out)
+        categories = [val.ljust(col_width[i]) for i, val in enumerate(category_head)]
+        print(",".join(categories), file=out)
+        for row in rows:
+            padded_row = [val.ljust(col_width[i]) for i, val in enumerate(row)]
+            print(",".join(padded_row), file=out)
+
+    print("Written:", output_file.resolve())
+
+
+def pin_int(pname):
```

Comment:
Please add a comment explaining what this does (e.g. just an example input and output)